### ISSUE TITLE
catalog:images:resize fails to process all images -> Possible underlying Magento/Framework/DB/Query/Generator issue

### ISSUE THREAD

**Issue by gwharton** [magento/magento2#18387]
### Preconditions
1. Magento 2.3-develop
2. Sample data deployed (in order to easily have several test images in store)

### Steps to reproduce
1. php bin/magento catalog:images:resize

### Expected result
1. All 3422 images should be processed

### Actual result
1. Only 801 images are processed.
```
developer@reedling$ php bin/magento catalog:images:resize
 801/3422 [======>---------------------]  23% 14 mins 58.0 MiB  | /w/s/wsh12-red_main_2.jpg
Product images resized successfully
developer@reedling$
```

In my 2.2.6 development store which has 140 images, it always stops after exactly 100.

I can reproduce this issue on both 2.3-develop and 2.2.6.

Possible underlying Magento/Framework/DB/Query/Generator issue.

**Comment by magento-engcom-team** [comment]
Hi @gwharton. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@gwharton do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by gwharton** [comment]
Can't use engcom test instances as need terminal and sample data.

However 100% reproducable using above steps on newly deployed 2.3-develop from git, and 2.2.6 existing installs (two 2.2.6 deployments both show same issue)

**Comment by Fazzke** [comment]
#squashtoberfest

**Comment by magento-engcom-team** [comment]
@Fazzke thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by magento-engcom-team** [comment]
Hi @Fazzke. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by Fazzke** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @Fazzke. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @Fazzke, here is your [Magento instance](https://i-18387-2-2-develop.engcom.dev.magento.com).
Admin access: [https://i-18387-2-2-develop.engcom.dev.magento.com/admin](https://i-18387-2-2-develop.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
Hi @vpodorozh. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by vpodorozh** [comment]
@gwharton - all your images are processed correctly.
However, there is a problem with total images count calculation select statement here: https://github.com/magento/magento2/pull/18807/commits/6a6079dcde8068dbac8173699ae4450079ed3201#diff-1c09f9f539ac83d47bf95f4f1ef90457R78
Distinct by image path is not used in count(*).

So, in case you will execute the next SQL:
`select count(DISTINCT value) from catalog_product_entity_media_gallery as g where disabled = 0;`
You will get "801" - the actual amount of your images.

I've fixed total images calculation my PR.

**Comment by gwharton** [comment]
Something doesnt sound right to me there.

I have 139 images in my pub/media/catalog/product folder
MYSQL says I have 139 images
But resize only resizes the first 100

```
developer@reedling:~/html$ php bin/magento catalog:images:resize                               
100/139 [====================>-------]  71% 59 secs 64.0 MiB    | /s/h/shell_omala_s2_gx_100.jpg
Product images resized successfully.
developer@reedling:~/html$ find pub/media/catalog -name *.jpg | grep -v cache | grep -v resized | grep -v placeholder | wc -l
139
developer@reedling:~/html$ mysql dev_magento
mysql> select count(DISTINCT value) from catalog_product_entity_media_gallery as g where disabled = 0;
+-----------------------+
| count(DISTINCT value) |
+-----------------------+
|                   139 |
+-----------------------+
1 row in set (0.00 sec)

mysql> 
```

**Comment by gwharton** [comment]
Seems to be two issues here as if I do it on a clean 2.2-develop with sample data installed.
```
developer@reedling:~/dev1$ php bin/magento catalog:images:resize                                                         
 700/3422 [=====>----------------------]  20% 15 mins 54.0 MiB  | /w/s/wsh07-blue_main.jpg
Product images resized successfully.
developer@reedling:~/dev1$ find -L pub/media/catalog -name *.jpg | grep -v cache | grep -v resized | grep -v placeholder | wc -l
725
developer@reedling:~/dev1$ mysql dev1_magento
mysql> select count(DISTINCT value) from catalog_product_entity_media_gallery as g where disabled = 0;
+-----------------------+
| count(DISTINCT value) |
+-----------------------+
|                   723 |
+-----------------------+
1 row in set (0.00 sec)

mysql> 
```
I'm willing to accept that there are two images in the pub/media folder that don't correspond to an entry in the database which exlpains the discrepancy between 723 and 725. 723 being the number of images correctly configured.

The images resize command wrongly calculates the total as 3422 which looks like the issue you are fixing with your PR's.

The image resize command also stops at 700 which is a separate bug. This second problem seems to be related to the batching with the batchSelectIterator inside the getAllProductImages() call of /app/code/Magento/Catalog/Model/ResourceModel/Product/Image.php

The default batch size is 100, and it looks like it doesnt process the last batch.

So in my first example it processes the first 100, then skips the last 39. And in my second example it processes the first (7*100) and then skips the last 23.

**Comment by gwharton** [comment]
Perhaps this second bug is not present in 2.3-develop as if you run the same on that, you get 801 images processed of 801 images configured (albeit reporting that there are 3422 images in total (total fixed by your PR))

**Comment by vpodorozh** [comment]
@gwharton - thank you for your response and explanations.
I've developed the fix for 2.3-develop branch so I've not reproduced the second problem, which is actually critical.
WIll take a look at 2.2-develop branch on the next week.

**Comment by vpodorozh** [comment]
@gwharton - you were totally right! Bug is present in 2.2-develop branch and the reason in `\Magento\Framework\DB\Query\BatchRangeIterator`.
This issue #12624 has fixed this problem in `2.3-develop` branch, however, backport to `2.2-develop` branch was not made. So I've made it in the scope of current PR https://github.com/magento/magento2/pull/18809 - you can check the fix here https://github.com/magento/magento2/pull/18809/commits/76bd08924e7e0f8a3fb88b3990623e5872b10760

**Comment by sidolov** [comment]
Hi @gwharton. Thank you for your report. 
The issue has been fixed in magento/magento2#18807 by @vpodorozh in 2.3-develop branch
Related commit(s):
 - [6a6079dcde8068dbac8173699ae4450079ed3201](https://github.com/magento/magento2/commit/6a6079dcde8068dbac8173699ae4450079ed3201)
 - [466daaa332e85eee15d3391d0ceead11855e90dd](https://github.com/magento/magento2/commit/466daaa332e85eee15d3391d0ceead11855e90dd)
 - [e1a1c7b280bfbb6c8cbcaf205d0d7f7dbb882878](https://github.com/magento/magento2/commit/e1a1c7b280bfbb6c8cbcaf205d0d7f7dbb882878)
 - [f6b0a2abfa0c95ce406a707a3ede42dbad5cfc22](https://github.com/magento/magento2/commit/f6b0a2abfa0c95ce406a707a3ede42dbad5cfc22)
 - [9e9715ba7d19051a87ba91cfdeac5bbd6c7b3d8e](https://github.com/magento/magento2/commit/9e9715ba7d19051a87ba91cfdeac5bbd6c7b3d8e)
 - [546e7ce1c64fa8f0290eabd26070eac4837cb39e](https://github.com/magento/magento2/commit/546e7ce1c64fa8f0290eabd26070eac4837cb39e)
 - [f38a4fc3e78208a8d63e545bffe81c444a1e09a8](https://github.com/magento/magento2/commit/f38a4fc3e78208a8d63e545bffe81c444a1e09a8)
 - [12f839644f71915e165a2804f505a727cd477cfc](https://github.com/magento/magento2/commit/12f839644f71915e165a2804f505a727cd477cfc)
 - [c95ce3ca53529c1a1e38f2ec5b187c64f45029e6](https://github.com/magento/magento2/commit/c95ce3ca53529c1a1e38f2ec5b187c64f45029e6)
 - [c1aa5bc9c6c6d69c399a1fe4070c508fbcb94a8b](https://github.com/magento/magento2/commit/c1aa5bc9c6c6d69c399a1fe4070c508fbcb94a8b)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @gwharton , thank you for you report, this issue has already fixed in 2.3-develop branch, and will be available on 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @gwharton. Thank you for your report. 
The issue has been fixed in magento/magento2#18809 by @vpodorozh in 2.2-develop branch
Related commit(s):
 - [f114f2aa61decaaecff02ebc6e2cabcc12569bd3](https://github.com/magento/magento2/commit/f114f2aa61decaaecff02ebc6e2cabcc12569bd3)
 - [95eeb9f6a022b1195b838965f65486a503d57137](https://github.com/magento/magento2/commit/95eeb9f6a022b1195b838965f65486a503d57137)
 - [84590ded5ed62edede5a12a68d384433e51ce0f3](https://github.com/magento/magento2/commit/84590ded5ed62edede5a12a68d384433e51ce0f3)
 - [4653d3e007500133b1b55012b469fd782738c680](https://github.com/magento/magento2/commit/4653d3e007500133b1b55012b469fd782738c680)
 - [d2e53455f4a79a3dc71e9e267259bfabc799a577](https://github.com/magento/magento2/commit/d2e53455f4a79a3dc71e9e267259bfabc799a577)
 - [2539a716f5d54c66c89f64965d6fb6762b6c90e8](https://github.com/magento/magento2/commit/2539a716f5d54c66c89f64965d6fb6762b6c90e8)
 - [76bd08924e7e0f8a3fb88b3990623e5872b10760](https://github.com/magento/magento2/commit/76bd08924e7e0f8a3fb88b3990623e5872b10760)
 - [b81b5503a676778793eac3a321d59fd3741935ca](https://github.com/magento/magento2/commit/b81b5503a676778793eac3a321d59fd3741935ca)
 - [051d82db553171cd784f543747ccff926287605d](https://github.com/magento/magento2/commit/051d82db553171cd784f543747ccff926287605d)
 - [10a1a881e8c7cbf983b800fe6479127e159a86e2](https://github.com/magento/magento2/commit/10a1a881e8c7cbf983b800fe6479127e159a86e2)

The fix will be available with the upcoming 2.2.8 release.
